pkg/controller/node/node_controller.go

```
@@ -128,6 +143,199 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
+			glog.Errorf("Acceptable values for percentangeOfMastersSchedulable are 100 and 0 but found %v", desiredSchedulableMasters)
+			return
+		}
+	case intstrutil.Int:
+		if desiredSchedulableMasters.IntValue() == 0 {
+			desiredSchedulableMasters = intstrutil.FromInt(0)
+		}
+		// If this value is greater than no of master nodes, so set it to all the master available.
+		if desiredSchedulableMasters.IntValue() <= currentMasterNodesCount || desiredSchedulableMasters.IntValue() > currentMasterNodesCount {
+			desiredSchedulableMasters = intstrutil.FromInt(currentMasterNodesCount)
+		}
+	}
+	return
+}
+
+// makeMasterNodesSchedulable takes currentUnSchedulableMasters and number of master nodes which need to be made
+// schedulable as requested by user. As of now, we take unschedulable master nodes slice and iterate over it to make
+// them schedulable. As of now, I don't see a reason to randomize it.
+func (crtl *Controller) makeMasterNodesSchedulable(currentUnSchedulableMasters []*corev1.Node, neededSchedulableMasters int) {
+	for i := 0; i < neededSchedulableMasters; i++ {
+		// TODO: Add a condition when master node is in updating stage?
+		_, err := updateNodeRetry(crtl.kubeClient.CoreV1().Nodes(), crtl.nodeLister, currentUnSchedulableMasters[i].Name, func(node *corev1.Node) {
+			// Add worker label
+			newLabels := node.Labels
+			if _, hasWorkerLabel := newLabels["node-role.kubernetes.io/worker"]; !hasWorkerLabel {
+				newLabels["node-role.kubernetes.io/worker"] = ""
+			}
+			node.Labels = newLabels
+			// Remove master taint
+			newTaints := []corev1.Taint{}
+			for _, t := range node.Spec.Taints {
+				if t.Key == "node-role.kubernetes.io/master" && t.Effect == corev1.TaintEffectNoSchedule {
+					continue
+				}
+				newTaints = append(newTaints, t)
+			}
+			//
+			node.Spec.Taints = newTaints
+		})
+		if err != nil {
+			glog.Errorf("error making node %v schedulable with error %v", err)
```

As a general rule, rather than logging and continuing, please return the error and let the higher level logic retry.

---

In this case, it might be different - we cannot return error when it fails for a single node, perhaps we need to capture slice of errors for individual nodes and return. I don't have a strong preference but I see your point.

---

> we cannot return error when it fails for a single node, 

Why not?  

> perhaps we need to capture slice of errors for individual nodes and return.

Yes, that's a better pattern indeed, but I think "fast fail for single item" is still better than "log and continue".

---

Yeah, I have done slice collection and then returned because in the next iteration it could cause problems with other node. Let me know, if you prefer it other way